stdlib/src/utils/span.mojo

```diff
@@ -335,3 +338,194 @@ struct Span[
         return Span[T, _lit_mut_cast[lifetime, False].result](
             unsafe_ptr=self._data, len=self._len
         )
+
+    fn find[
```

Comment:
**Question** It seems a bit weird from an API design perspective to have this "find" function in `span`, can you help me understand why we'd want this?  It feels more coupled to string algorithms, and `Span` isn't just for a view over string data.

Reply:
It can be used if you want to find occurrences of a specific scalar value sequence. My main thought was that we'll lose the ability to do that once `StringSlice.find()` is fixed to use unicode codepoints and it introduces quite some overhead for algorithms that are faster using raw bytes. We could also make this private WDYT?

Reply:
@JoeLoser gentle ping and also to add something, since it seems like we are moving toward unifying Python's `bytes` into `Span[UInt8]`in PR #3636, request #3634 and many other places. Python's [bytes.find](https://docs.python.org/3/library/stdtypes.html#bytes.find) is actually a method, and many other things that we put in `StringSlice` and `StringRef` are actually [operations from `bytes`](https://docs.python.org/3/library/stdtypes.html#bytes-and-bytearray-operations).

Reply:
I think this is a great change for overall performance when manipulating slices. It opens a lot of possibilities.

Reply:
@JoeLoser per discussion on discord, `Span[T].find` is intended to be `memchr` or `memmem` as appropriate, but that the subspan of bytes must be aligned to `alignof[T]()`. We can probably expand it to any `T: Trivial` once we have a `Trivial` trait. 

The proper string version likely needs be specialized for UTF-8 comparisons. @martinvuyk and I have discussed an "AsciiString" which can more aggressively optimize for using byte-oriented intrinsics since UTF-8 is annoying to do with SIMD.